include/fem-dem/void_fraction.h

```diff
@@ -0,0 +1,404 @@
+  /**
+   * @brief Setup the degrees of freedom
+   *
+   */
+  void
+  setup_dofs() override;
+
+
+  /**
+   * @brief Establishes the constraints of the void fraction systems.
+   *
+   */
+  void
+  setup_constraints(
+    const BoundaryConditions::NSBoundaryConditions<dim> &boundary_conditions);
+
+
+  /**
+   * @brief Assembles the diagonal of the mass matrix to impose constraints on the system
+   *
+   *  @param diagonal_mass_matrix The matrix for which the diagonal entries will be filled with the diagonal of the mass matrix
+   *
+   *  @todo Establish if it is really necessary to keep this as a matrix and not as a vector since a diagonal is nothing more than a vector.
+   */
+  void
+  assemble_mass_matrix_diagonal(
+    TrilinosWrappers::SparseMatrix &diagonal_mass_matrix);
+
+  /**
+   * @brief Calculates the void fraction
+   *
+   * @param time current time for which the void fraction is to be calculated.
+   *
+   */
+  void
+  calculate_void_fraction(const double time);
+
+
+  /**
+   * @brief Assemble and solve the system.
+   *
+   * @param[in] is_post_mesh_adaptation Indicates if the equation is being
+   * solved during post_mesh_adaptation(), for verbosity.
+   */
+  void
+  solve(const bool & /*is_post_mesh_adaptation*/) override
+  {}
+
+
+  /**
+   * @brief Percolates the time vector for the void fraction. This operation is called at the end of a time-step.
+   *
+   */
+  void
+  percolate_void_fraction()
+  {
+    for (unsigned int i = previous_void_fraction.size() - 1; i > 0; --i)
+      {
+        previous_void_fraction[i] = previous_void_fraction[i - 1];
+      }
+    previous_void_fraction[0] = void_fraction_locally_relevant;
+  }
+
+
+  /**
+   * @brief Initializes the void fraction at the begginig of a simulation
+   *
+   */
+
+  void
+  initialize_void_fraction(const double time)
+  {
+    calculate_void_fraction(time);
+    for (auto &previous_solution : this->previous_void_fraction)
+      previous_solution = void_fraction_locally_relevant;
+  }
+
+  /// DoFHandler that manages the void fraction
+  DoFHandler<dim> dof_handler;
+
+  /// The solutions are made public instead of using getters
+  /// Solution of the void fraction at previous time steps
+  std::vector<GlobalVectorType> previous_void_fraction;
+
+  /// Fully distributed (including locally relevant) solution
+  GlobalVectorType void_fraction_locally_relevant;
+
+  /// Finite element for the void fraction
+  std::shared_ptr<FiniteElement<dim>> fe;
+
+  /// Index set for the locally owned degree of freedoms
+  IndexSet locally_owned_dofs;
+
+  /// Index set for the locally relevant degree of freedoms
+  IndexSet locally_relevant_dofs;
+
+  /// Constraints used for the boundary conditions of the void fraction.
+  /// Currently, this is only used to establish periodic void fractions. This
+  /// object has to be made public because the boundary conditions are set
+  /// outside of the object for now.
+  AffineConstraints<double> void_fraction_constraints;
```

Comment:
```suggestion
  // DoFHandler that manages the void fraction
  DoFHandler<dim> dof_handler;

  // The solutions are made public instead of using getters
  // Solution of the void fraction at previous time steps
  std::vector<GlobalVectorType> previous_void_fraction;

  // Fully distributed (including locally relevant) solution
  GlobalVectorType void_fraction_locally_relevant;

  // Finite element for the void fraction
  std::shared_ptr<FiniteElement<dim>> fe;

  // Index set for the locally owned degree of freedoms
  IndexSet locally_owned_dofs;

  // Index set for the locally relevant degree of freedoms
  IndexSet locally_relevant_dofs;

  // Constraints used for the boundary conditions of the void fraction.
  // Currently, this is only used to establish periodic void fractions. This
  // object has to be made public because the boundary conditions are set
  // outside of the object for now.
  AffineConstraints<double> void_fraction_constraints;
```

Reply:
I think the /// is necessary for DOXYGEN